perception/sub8_batcave_ogrid/src/OGridGen.cpp

```diff
@@ -0,0 +1,133 @@
+#include "OGridGen.hpp"
+
+//TODO: Add service call to clear ogrid
+
+OGridGen::OGridGen() : nh(ros::this_node::getName()), classification(&nh)
+{
+  pubGrid = nh.advertise<nav_msgs::OccupancyGrid>("/ogrid", 10, true);
+  pubPointCloud = nh.advertise<sensor_msgs::PointCloud2>("/point_cloud", 1);
+  nh.param<float>("resolution", resolution, 0.2f);
+  nh.param<float>("ogrid_size", ogrid_size, 50.f);
+  nh.param<float>("pool_depth", pool_depth, 7.f);
+  timer = nh.createTimer(ros::Duration(0.3), std::bind(&OGridGen::publish_ogrid, this, std::placeholders::_1));
+  matOgrid = cv::Mat::zeros((int)(ogrid_size / resolution), (int)(ogrid_size / resolution), CV_8U);
+
+  get_bounds = nh.serviceClient<sub8_msgs::Bounds>("get_bounds");
+
+  sub = nh.subscribe("/blueview_driver/ranges", 1, &OGridGen::callback, this);
+
+  int pointCloudBufferSize;
+  nh.param<int>("buffer_size", pointCloudBufferSize, 5000);
+  pointCloudBuffer.resize(pointCloudBufferSize);
+}
+
+void OGridGen::publish_ogrid(const ros::TimerEvent &)
+{
+  // Draw ogrid
+  sub8_msgs::Bounds bound_hold;
+  if (get_bounds.call(bound_hold))
+  {
+    bounds = {};
+    for (auto &p : bound_hold.response.bounds)
+    {
+      bounds.push_back(cv::Point(p.x / resolution + matOgrid.cols / 2, p.y / resolution + matOgrid.rows / 2));
+    }
+    const cv::Point *pts = (const cv::Point *)cv::Mat(bounds).data;
+    int npts = cv::Mat(bounds).rows;
+    cv::polylines(matOgrid, &pts, &npts, 1, true, 255, 3, CV_8U, 0);
+  }
+
+  // Flatten Ogrid
+  nav_msgs::OccupancyGrid rosGrid;
+  std::vector<int8_t> data;
+  for (int i = 0; i < matOgrid.cols; ++i)
+  {
+    for (int j = 0; j < matOgrid.rows; j++)
+    {
+      data.push_back((int)matOgrid.at<uint8_t>(cv::Point(i, j)));
+    }
+  }
+
+  // Publish Ogrid
+  rosGrid.header.seq = 0;
+  rosGrid.info.resolution = resolution;
+  rosGrid.header.frame_id = "map";
+  rosGrid.header.stamp = ros::Time::now();
+  rosGrid.info.map_load_time = ros::Time::now();
+  rosGrid.info.width = matOgrid.cols;
+  rosGrid.info.height = matOgrid.rows;
+  rosGrid.info.origin.position.x = -ogrid_size / 2;
+  rosGrid.info.origin.position.y = -ogrid_size / 2;
+  rosGrid.info.origin.position.z = 0;
+  rosGrid.data = data;
+  pubGrid.publish(rosGrid);
```

Comment:
Does this produce a flipped ogrid? I remember doing this in numpy would always result in a flipped/upside down ogrid from what I was expecting